File: android/app/src/main/java/gg/yappy/app/ui/components/Flair.kt
package gg.yappy.app.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawOutline
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.graphics.drawscope.translate
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import gg.yappy.app.data.ConversationAppearance

/**
 * Group flair rendering.
 *
 * A group's flair is its owner's one visual sentence about the place, so it has
 * to be *visible* (the whole point is the conversation list) without breaking
 * the design system. The rule: flair may add colour and motion *around* an
 * element — the ring, a glow, a tinted title — but never restyles the element
 * itself. That is what keeps a flaired row and a plain row siblings instead of
 * strangers.
 */

fun flairColor(hex: String?): Color? = hex?.let {
    runCatching { Color(android.graphics.Color.parseColor(it)) }.getOrNull()
}

/**
 * Paints whatever the element draws (text, icons) with [brush] via an SrcIn
 * blend — the portable way to get gradient text without depending on
 * TextStyle's brush support. The 0.99 alpha forces an offscreen layer, which
 * the blend needs to composite against.
 */
fun Modifier.gradientFill(brush: Brush): Modifier = this
    .graphicsLayer(alpha = 0.99f)
    .drawWithContent {
        drawContent()
        drawRect(brush = brush, blendMode = BlendMode.SrcIn)
    }

/** The two ring stops: explicit gradient, else accent doubled, else null. */
fun ConversationAppearance.ringColors(): List<Color>? {
    val fromGradient = gradient?.mapNotNull { flairColor(it) }?.takeIf { it.size >= 2 }
    if (fromGradient != null) return fromGradient
    val solid = flairColor(accent) ?: return null
    return listOf(solid, solid.copy(alpha = 0.55f))
}

/** Colour a flaired title should take: the accent, else the first gradient stop. */
fun ConversationAppearance.titleColor(): Color? =
    flairColor(accent) ?: gradient?.firstOrNull()?.let { flairColor(it) }

/**
 * An [Avatar] wrapped in its group's flair: a gradient ring, optionally glowing
 * or shimmering. Falls back to a plain avatar when there is nothing to show, so
 * call sites do not need to branch.
 */
@Composable
fun FlairAvatar(
    appearance: ConversationAppearance?,
    url: String?,
    name: String?,
    id: String,
    modifier: Modifier = Modifier,
    size: Dp = 48.dp,
    presence: String? = null,
    /** Squircle for groups, circle for people — see [gg.yappy.app.ui.theme.PlaceShape]. */
    shape: androidx.compose.ui.graphics.Shape = androidx.compose.foundation.shape.CircleShape,
) {
    val ring = appearance?.ringColors()
    if (ring == null) {
        Avatar(url, name, id, modifier, size, presence, shape)
        return
    }

    // Shimmer rotates the gradient around the ring; a static ring is angle 0.
    // One infinite transition per avatar is fine — the list shows tens, not
    // thousands, and Compose elides the animation when nothing reads it.
    val angle = if (appearance.effect == "shimmer") {
        val transition = rememberInfiniteTransition(label = "flair-shimmer")
        val value by transition.animateFloat(
            initialValue = 0f,
            targetValue = 360f,
            animationSpec = infiniteRepeatable(tween(2600, easing = LinearEasing)),
            label = "flair-angle",
        )
        value
    } else 0f

    val stroke = 2.5.dp
    val gap = 2.5.dp
    val glow = appearance.effect == "glow"

    // The ring is drawn *inside* the declared size and the avatar shrinks to
    // fit, so flaired and plain rows keep identical layout metrics.
    Box(
        modifier = modifier
            .size(size)
            .drawBehind {
                val radius = this.size.minDimension / 2f
                if (glow) {
                    drawCircle(
                        brush = Brush.radialGradient(
                            colors = listOf(ring.first().copy(alpha = 0.40f), Color.Transparent),
                            center = Offset(this.size.width / 2f, this.size.height / 2f),
                            radius = radius * 1.35f,
                        ),
                        radius = radius * 1.35f,
                    )
                }
                // The ring follows the avatar's silhouette — a circle for a
                // person, a squircle for a place. Inset by half the stroke so
                // the ring stays inside the declared bounds.
                val strokePx = stroke.toPx()
                val ringSize = androidx.compose.ui.geometry.Size(
                    this.size.width - strokePx,
                    this.size.height - strokePx,
                )
                rotate(angle) {
                    translate(strokePx / 2f, strokePx / 2f) {
                        drawOutline(
                            outline = shape.createOutline(ringSize, layoutDirection, this),
                            brush = Brush.sweepGradient(ring + ring.first()),
                            style = Stroke(width = strokePx),
                        )
                    }
                }
            }
            .padding(stroke + gap),
        contentAlignment = Alignment.Center,
    ) {
        Avatar(url, name, id, size = size - (stroke + gap) * 2, presence = presence, shape = shape)
    }
}

File: ios/Yappy/Components/Flair.swift
import SwiftUI

/// Group flair rendering.
///
/// A group's flair is its owner's one visual sentence about the place, so it has
/// to be *visible* (the whole point is the conversation list) without breaking
/// the design system. The rule: flair may add colour and motion *around* an
/// element — the ring, a glow, a tinted title — but never restyles the element
/// itself. That is what keeps a flaired row and a plain row siblings instead of
/// strangers.
extension ConversationAppearance {
    /// The two ring stops: explicit gradient, else accent doubled, else nil.
    var ringColors: [Color]? {
        if let gradient {
            let parsed = gradient.compactMap { Color(hexString: $0) }
            if parsed.count >= 2 { return parsed }
        }
        guard let solid = Color(hexString: accent) else { return nil }
        return [solid, solid.opacity(0.55)]
    }

    /// Colour a flaired title should take: the accent, else the first gradient
    /// stop.
    var titleColor: Color? {
        Color(hexString: accent) ?? gradient?.first.flatMap { Color(hexString: $0) }
    }

    /// A gradient an outgoing bubble or a send button can wear. Nil unless the
    /// group actually declared two stops — one colour is not a gradient, and
    /// faking the second makes every flaired group look the same.
    var bubbleGradient: LinearGradient? {
        guard let gradient else { return nil }
        let parsed = gradient.compactMap { Color(hexString: $0) }
        guard parsed.count >= 2 else { return nil }
        return LinearGradient(colors: parsed, startPoint: .topLeading, endPoint: .bottomTrailing)
    }
}

/// An `Avatar` wrapped in its group's flair: a gradient ring, optionally glowing
/// or shimmering. Falls back to a plain avatar when there is nothing to show, so
/// call sites do not need to branch.
struct FlairAvatar: View {
    let appearance: ConversationAppearance?
    let url: String?
    let name: String?
    let id: String
    var size: CGFloat = 48
    var presence: String?
    var shape: AvatarShape = .person

    @State private var angle: Double = 0

    var body: some View {
        if let ring = appearance?.ringColors {
            flaired(ring)
        } else {
            Avatar(url: url, name: name, id: id, size: size, presence: presence, shape: shape)
        }
    }

    @ViewBuilder
    private func flaired(_ ring: [Color]) -> some View {
        let stroke: CGFloat = 2.5
        let gap: CGFloat = 2.5
        let inner = size - (stroke + gap) * 2
        let glow = appearance?.effect == "glow"
        let shimmer = appearance?.effect == "shimmer"

        ZStack {
            if glow {
                Circle()
                    .fill(
                        RadialGradient(
                            colors: [ring[0].opacity(0.40), .clear],
                            center: .center,
                            startRadius: 0,
                            endRadius: size * 0.675
                        )
                    )
                    .frame(width: size * 1.35, height: size * 1.35)
            }

            // The ring follows the avatar's silhouette — a circle for a person, a
            // squircle for a place — and is drawn *inside* the declared size so
            // flaired and plain rows keep identical layout metrics.
            ringShape
                .stroke(
                    AngularGradient(colors: ring + [ring[0]], center: .center),
                    lineWidth: stroke
                )
                .frame(width: size - stroke, height: size - stroke)
                .rotationEffect(.degrees(angle))

            Avatar(url: url, name: name, id: id, size: inner, presence: presence, shape: shape)
        }
        .frame(width: size, height: size)
        .onAppear {
            // Shimmer rotates the gradient around the ring; a static ring is
            // angle 0 and never starts a timer.
            guard shimmer else { return }
            withAnimation(.linear(duration: 2.6).repeatForever(autoreverses: false)) {
                angle = 360
            }
        }
    }

    private var ringShape: AnyShape {
        switch shape {
        case .person: return AnyShape(Circle())
        case .place: return AnyShape(PlaceShape())
        }
    }
}

/// Paints whatever the view draws — text, a mark — with a gradient.
///
/// `foregroundStyle` takes a `ShapeStyle` directly on iOS 17, so unlike the
/// Compose original this needs no offscreen layer and no blend mode.
extension View {
    func gradientFill(_ gradient: LinearGradient) -> some View {
        overlay(gradient).mask(self)
    }
}

/// The brand gradient: the accent running to teal. Used only where the app is
/// talking about itself — the sign-in mark and the home lockup.
func brandGradient(_ colors: NeuColors) -> LinearGradient {
    LinearGradient(
        colors: [colors.accent, Color(hex: 0x00CEC9)],
        startPoint: .leading,
        endPoint: .trailing
    )
}
